### PR TITLE
Require `then` and `else` with `if`

### PR DESCRIPTION
**Pull request by tlively**
We previously supported (and primarily used) a non-standard text format for
conditionals in which the condition, if-true expression, and if-false expression
were all simply s-expression children of the `if` expression. The standard text
format, however, requires the use of `then` and `else` forms to introduce the
if-true and if-false arms of the conditional. Update the legacy text parser to
require the standard format and update all tests to match. Update the printer to
print the standard format as well.

The .wast and .wat test inputs were mechanically updated with this script:
https://gist.github.com/tlively/85ae7f01f92f772241ec994c840ccbb1

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #6201** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6201?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/6201?utm_source=stack-comment).

**Comment by tlively** [comment]
I strongly recommend hiding whitespace when viewing the diff 😆

File: test/gtest/cfg.cpp [hunk 1]
```diff
@@ -65,7 +67,7 @@ TEST_F(CFGTest, Print) {
 ;; preds: [3], succs: [6]
 4:
   6: i32.const 3
-  7: block
+  7: block (result i32)
```
Comment (kripken) [inline test/gtest/cfg.cpp:70]: Unrelated change?

File: test/wasm2js/fac.2asm.js [hunk 2]
```diff
@@ -17,67 +17,61 @@ function asmFunc(imports) {
...
+  var i64toi32_i32$3 = 0, i64toi32_i32$5 = 0, i64toi32_i32$1 = 0, $8 = 0, $8$hi = 0, $6 = 0, $6$hi = 0;
   i64toi32_i32$1 = 0;
   i64toi32_i32$3 = 0;
-  if ((i64toi32_i32$2 | 0) == (i64toi32_i32$3 | 0) & ($0$hi | 0) == (i64toi32_i32$1 | 0) | 0) {
-   i64toi32_i32$2 = 0;
-   $12 = 1;
-   $12$hi = i64toi32_i32$2;
+  if (($0_1 | 0) == (i64toi32_i32$3 | 0) & ($0$hi | 0) == (i64toi32_i32$1 | 0) | 0) {
+   $8 = 1;
+   $8$hi = 0;
```
Comment (kripken) [inline test/wasm2js/fac.2asm.js:25]: I am surprised to see codegen changes here. This change even looks non-NFC. Do you know why this changed?

**Review by kripken — COMMENTED** [review]
Ah, this must have changed when wasm switched to a stack machine... because it is needed to identify structural children.

lgtm % questions

File: test/gtest/cfg.cpp (hunk 1 above)
Comment (tlively) [reply test/gtest/cfg.cpp:70]: No, this is a necessary change. Probably due to a difference in finalization between the parsing for `block` and the parsing for `then`, but not sure it's worth digging into more.

Comment (kripken) [reply test/gtest/cfg.cpp:70]: Makes sense, it's probably that.

File: test/wasm2js/fac.2asm.js (hunk 2 above)
Comment (tlively) [reply test/wasm2js/fac.2asm.js:25]: It looks like the source file contains `then` and `else` blocks containing single instructions. Those were previously parsed as singleton blocks, but now they're parsed as the actual instructions. I think that change to the IR probably explains the different output here.

Comment (kripken) [reply test/wasm2js/fac.2asm.js:25]: Could be. But looking at the output it's not obvious to me if it is or isn't a regression (line 21 is gone, for example - these aren't just renamed variables, there is a structural change), so I think it's worth investigating. Let me know if you'd like help with that.

Comment (tlively) [reply test/wasm2js/fac.2asm.js:25]: Ok, I verified that the different parsing behavior explains this. I ran wasm2js on one of the source functions with and without the special case for singleton `then` and `else` blocks and the output had precisely these differences.

Comment (kripken) [reply test/wasm2js/fac.2asm.js:25]: Great, thanks!

**Review by kripken — APPROVED** [review]

**Comment by tlively** [comment]
### Merge activity

* **Jan 4, 5:25 PM**: @tlively started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/6201).

**Comment by kripken** [comment]
I see a fuzzer error shortly after this landed which bisects down to this:
```
running testcase handler: Wasm2JS                                                              
wasm-opt out/test/a.wasm --legalize-js-interface -o /usr/local/google/home/azakai/Dev/2-binaryen
/out/test/a.wasm.temp.wasm                                                                                                                                                                    
wasm-opt out/test/a.wasm.temp.wasm -o out/test/a.wa
sm.temp.wasm --stub-unsupported-js --dealign --alignment-lowering
wasm-opt out/test/a.wasm.temp.wasm -o out/test/b.wa
sm.temp.wasm --roundtrip -O1 --optimize-casts --code-pushing --simplify-locals-notee
wasm-opt out/test/a.wasm.temp.wasm --emit-js-wrapper=/dev/stdout
warning: no passes specified, not doing any work
warning: no output file specified, not emitting output
wasm2js out/test/a.wasm.temp.wasm --emscripten -O --deterministic
terminate called after throwing an instance of 'wasm::SParseException'                                                                                                                        
```
That is on seed `10441444793052190197`.

Reduced testcase:
```wat
(module
 (type $0 (func (result i32)))
 (func $0 (result i32)
  (i32.popcnt
   (i32.const 0)
  )
 )
)
```
STR:
```
$ bin/wasm-as w.wat -o a.wasm
$ bin/wasm2js a.wasm
terminate called after throwing an instance of 'wasm::SParseException'
Aborted
```
Oddly it's not a text format issue, as it happens when parsing a binary..?

@tlively
